src/uucore/src/lib/mods/error.rs

```diff
@@ -221,13 +235,130 @@ impl Display for UError {
+///     // Perform computations here ...
+///     return Err(LsError::InvalidLineWidth(String::from("test")).into())
+/// }
+/// ```
+///
+/// The call to `into()` is required to convert the [`UCustomError`] to an
+/// instance of [`UError`].
+///
+/// A crate like [`quick_error`](https://crates.io/crates/quick-error) might
+/// also be used, but will still require an `impl` for the `code` method.
 pub trait UCustomError: Error {
+    /// Error code of a custom error.
+    ///
+    /// Implement this for each variant of an enum-type to associate an error
+    /// code (which is returned to the system shell) with an error.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use uucore::error::{UCustomError};
+    /// use std::{
+    ///     error::Error,
+    ///     fmt::{Display, Debug},
+    ///     path::PathBuf
+    /// };
+    ///
+    /// #[derive(Debug)]
+    /// enum MyError {
+    ///     Foo(String),
+    ///     Bar(PathBuf),
+    ///     Bing(),
+    /// }
+    ///
+    /// impl UCustomError for MyError {
+    ///     fn code(&self) -> i32 {
+    ///         match self {
+    ///             MyError::Foo(_) => 2,
+    ///             // All other errors yield the same error code, there's no
+    ///             // need to list them explicitly.
+    ///             _ => 1,
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// impl Error for MyError {}
+    ///
+    /// impl Display for MyError {
+    ///     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    ///         use MyError as ME;
+    ///         match self {
+    ///             ME::Foo(s) => write!(f, "Unknown Foo: '{}'", s),
+    ///             ME::Bar(p) => write!(f, "Couldn't find Bar: '{}'", p.display()),
+    ///             ME::Bing() => write!(f, "Exterminate!"),
+    ///         }
+    ///     }
+    /// }
+    /// ```
     fn code(&self) -> i32 {
         1
     }
 
+    /// Print usage help to a custom error.
+    ///
+    /// Implement this for each variant of an enum-type to control whether a
+    /// short usage help is printed below the error message. The usage help is
+    /// in the format: "Try '{name} --help' for more information." and printed
+    /// only for enum variants that return `true`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use uucore::error::{UCustomError};
+    /// use std::{
+    ///     error::Error,
+    ///     fmt::{Display, Debug},
+    ///     path::PathBuf
+    /// };
+    ///
+    /// #[derive(Debug)]
+    /// enum MyError {
+    ///     Foo(String),
+    ///     Bar(PathBuf),
+    ///     Bing(),
+    /// }
+    ///
+    /// impl UCustomError for MyError {
+    ///     fn code(&self) -> i32 {
+    ///         match self {
+    ///             MyError::Foo(_) => 2,
+    ///             // All other errors yield the same error code, there's no
+    ///             // need to list them explicitly.
+    ///             _ => 1,
+    ///         }
+    ///     }
+    ///
```

Comment:
`code` has a default implementation (defaulting to `1`), so it doesn't need to be in this example, I think

Reply:
That's right, but this is only the default for the whole enum type as long as I don't overload the function. It doesn't automatically apply the default to all variants which I don't cover. I tried it:

```
---- src/lib/mods/error.rs - mods::error::UCustomError::usage (line 314) stdout ----
error[E0004]: non-exhaustive patterns: `&Bar(_)` and `&Bing` not covered
  --> src/lib/mods/error.rs:331:15
   |
12 | / enum MyError {
13 | |     Foo(String),
14 | |     Bar(PathBuf),
   | |     --- not covered
15 | |     Bing(),
   | |     ---- not covered
16 | | }
   | |_- `MyError` defined here
...
20 |           match self {
   |                 ^^^^ patterns `&Bar(_)` and `&Bing` not covered
   |
   = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
   = note: the matched value is of type `&MyError`

```

Reply:
That's right. My idea was to remove the `code` function to simplify and focus the example. Just like the example for `code` leaves out the `usage` function.

Reply:
Sorry, didn't get that! You're right, it's not needed here. I removed it.